### PR TITLE
[DNM] Deathmatch

### PR DESCRIPTION
**Pull request by MadmanMartian**
Adds a deathmatch map, and sub systems that help such

 - Map Handler sub
 - - Itterates over all map handlers and processes them
- Map Handler
 - - Associated with a deathmatch map, holds the map spawners
 - - When a spawner's item is picked up, it registers itself to the map handler. Each time the map handler is ticked over, it ticks down the spawn timer. Once the timer hits 0, the spawner spawns a new item, and is removed from the map handler.
 - Item spawner
 - - When the map handler assumes control of a spawner, it forces it to spawn an item
 - - When a spawner spawns an item, it registers to the items pickup event
 - - When the item is picked up, the spawner will register itself to spawn a new item after so many seconds

Corrugated Observatory XX 2019-12-10 154240
Corrugated Observatory XX 2019-12-10 154227

Test maps:
 - fountain. 8 greytide spawns, 4 uzis, 4 explosive grenades, 2 energy shields, 2 ricochet rifles, 2 shotguns, 2 patient processors 
 - maint. 8 greytide spawns. 6 laser pistols, 2 NVGs, 4 bear traps, 2 shotguns, 2 patient processors

:cl:
 * rscadd: Adds some deathmatch maps

### CODE REVIEW

**Comment by ShiftyRail** [comment]
wait this isn't prison station

File: maps/misc/misc_map_items.dm [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	desc = "Will summon bloodthirsty murderers at random"
+	cooldown_duration = 10 SECONDS
+	print_path = /mob/living/carbon/human
+	var/obj/abstract/loadout/to_load = /obj/abstract/loadout/deathmatch
+
+/obj/machinery/mob_printer/deathmatch/finalize_spawn(var/mob/living/M)
+	new to_load(loc, M)
+	return M
+
+/obj/abstract/loadout/deathmatch
```
Comment (ShiftyRail) [inline maps/misc/misc_map_items.dm:12]: loadout datums hurt my soul

Comment (jknpj) [reply maps/misc/misc_map_items.dm:12]: outfit datums are still in the cuck branch

Comment (ShiftyRail) [reply maps/misc/misc_map_items.dm:12]: i know
i'll finish them eventually :tm:

**Comment by ghost** [comment]
SLAYER

**Comment by BomberBro** [comment]
dust2.dm when?

**Comment by calzilla1** [comment]
Wait, this isn't a cult/rev PR!

**Comment by brndd** [comment]
You might want to indicate the direction of the patient processors with decals or something.

**Comment by juanototo2** [comment]
Can you add this to deadchat kind of like bomberman is now?

**Comment by ghost** [comment]
Add a rocket launcher and a hectate please

**Comment by MadmanMartian** [comment]
No

**Comment by MadmanMartian** [comment]
pls no mergerino

File: code/controllers/subsystem/map_handlers.dm [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+			continue
+
+		M.Process()
+
+		if (MC_TICK_CHECK)
+			return
+
+/datum/map_handler
+	var/paused = FALSE
+	var/list/spawners = list()
```
Comment (PaulRitter) [inline code/controllers/subsystem/map_handlers.dm:39]: maybe call this `spawners_on_cooldown` and add a comment about how this list is only changed by the actual spawners themselves.

File: code/controllers/subsystem/map_handlers.dm [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+//If it has a respawn value left, we tick it down by 1
+//If it did have a respawn value left that was ticked down, and it is now 0, we handle it.
+/datum/map_handler/proc/Process()
+	for(var/obj/abstract/spawner/i in spawners)
+		if(spawners[i] > 0)
+			spawners[i]--
+			if(spawners[i] == 0)
+				handle_spawner(i)
+	if(--cleanup <= 0)
+		cleanup = initial(cleanup)
```
Comment (PaulRitter) [inline code/controllers/subsystem/map_handlers.dm:61]: make it a seperate var for adminbussing? maybe even make mob- and item-removal be toggable seperate from each other through flags?

File: code/controllers/subsystem/map_handlers.dm [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		for(var/obj/item/I in map)
+			var/obj/abstract/spawner/S = locate() in get_turf(I)
+			if(S && S.spawned == I)
+				continue //It has been spawned and not picked up
+			qdel(I)
```
Comment (PaulRitter) [inline code/controllers/subsystem/map_handlers.dm:66]: couldn't this in theory trigger right after an item has been spawned, blocking the spawner permanently?
shouldn't we therefore register our `S` in `spawners` / call `handle_spawner()` or `S.on_pickup()`?

File: code/controllers/subsystem/map_handlers.dm [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+			if(M.stat == DEAD)
+				qdel(M)
+
+/datum/map_handler/proc/handle_spawner(var/obj/abstract/spawner/I)
+	spawners[I] = null
+	spawners.Remove(I)
+	I.respawn()
+
+/obj/abstract/spawner
+	var/item_to_spawn
```
Comment (PaulRitter) [inline code/controllers/subsystem/map_handlers.dm:77]: a list perhaps?

**Comment by tuypo1** [comment]
make a multi z level 2fort

**Comment by tuypo1** [comment]
Actually if this does become a thing I might try my hand at making 2fort myself.
Or maybe that map from unreal tournament with the 2 big desert transports

**Comment by MadmanMartian** [comment]
sounds like fun!

**Comment by 9percents** [comment]
have you considered making some of walls on maint map as fake walls, windows or grills?
or maybe even a spawner of such walls?

**Comment by MadmanMartian** [comment]
I wanted de_maint to be claustrophobic, but fake walls does sound like a fun idea

**Comment by Kurfursten** [comment]
Reopen this if you decide to work on it again

**Pull request closed without merging**